myst_nbs/case_studies/wrapping_jax_function.myst.md

```diff
@@ -0,0 +1,765 @@
+    emission_signal = pm.Normal("emission_signal", 0, 1)
+    emission_noise = pm.HalfNormal("emission_noise", 1)
+
+    p_initial_state = pm.Dirichlet("p_initial_state", np.ones(3))
+    logp_initial_state = at.log(p_initial_state)
+
+    p_transition = pm.Dirichlet("p_transition", np.ones(3), size=3)
+    logp_transition = at.log(p_transition)
+
+    loglike = pm.Potential(
+        "hmm_loglike",
+        hmm_logp_op(
+            emission_observed,
+            emission_signal,
+            emission_noise,
+            logp_initial_state,
+            logp_transition,
+        ),
+    )
+```
+
+```{code-cell} ipython3
+---
+pycharm:
+  name: '#%%
+
+    '
+---
+pm.model_to_graphviz(model)
+```
+
+Before we start sampling, we check the logp of each variable at the model initial point. Bugs tend to manifest themselves in the form of `nan` or `-inf` for the initial probabilities.
+
+```{code-cell} ipython3
+initial_point = model.compute_initial_point()
+initial_point
+```
+
+```{code-cell} ipython3
+---
+pycharm:
+  name: '#%%
+
+    '
+---
+model.point_logps(initial_point)
+```
+
+We are now ready to sample!
+
+```{code-cell} ipython3
+---
+pycharm:
+  name: '#%%
+
+    '
+---
+with model:
+    idata = pm.sample(chains=2, cores=1)
+```
+
+```{code-cell} ipython3
+---
+pycharm:
+  name: '#%%
+
+    '
+---
+az.plot_trace(idata);
+```
+
+```{code-cell} ipython3
+true_values = [
+    emission_signal_true,
+    emission_noise_true,
+    *p_initial_state_true,
+    *p_transition_true.ravel(),
+]
+
+az.plot_posterior(idata, ref_val=true_values, grid=(3, 5));
+```
+
+The posteriors look reasonably centered around the true values used to generate our data.
+
++++
+
+## Unwrapping the wrapped JAX function
+
++++
+
+As mentioned in the beginning, Aesara can compile an entire graph to JAX. To do this, it needs to know how each {class}`~aesara.graph.op.Op` in the graph can be converted to a JAX function. This can be done by {term}`dispatch <dispatching>` with {func}`aesara.link.jax.dispatch.jax_funcify`. Most of the default Aesara {class}`~aesara.graph.op.Op`s already have such a dispatch function, but we will need to add a new one for our custom `HMMLogpOp`, as Aesara has never seen that before.
+
+For that we need a function which returns (another) JAX function, that performs the same computation as in our `perform` method. Fortunately, we started exactly with such function, so this amounts to 3 short lines of code.
+
+```{code-cell} ipython3
+@jax_funcify.register(HMMLogpOp)
+def hmm_logp_dispatch(op, **kwargs):
+    return vec_hmm_logp
+```
+
+:::{note}
+We do not return the jitted function, so that the entire Aesara graph can be jitted together after being converted to JAX.
+:::
+
+For a better understanding of {class}`~aesara.graph.op.Op` JAX conversions, we recommend reading Aesara's [Adding JAX and Numba support for Ops guide](https://aesara.readthedocs.io/en/latest/extending/creating_a_numba_jax_op.html?highlight=JAX)
```

Comment:
the link to the aesara guide should be `` {doc}`aesara:extending/creating_a_numba_jax_op` `` (it has no target so the doc way must be used)

Reply:
Thanks, updated

Reply:
Should I use something like this this for the first reference to Aesara as well? I did `[Aesara](https://aesara.readthedocs.io/en/latest/) `

Reply:
better to do it, then we can add a check against urls containing `aesara.readthedocs.io` more easily (like we did on pymc versioned docs). Link to the home page is `` {doc}`Aesara <aesara:index>` ``.

Reply:
Done